internal/cloud/realm/export.go

```diff
@@ -0,0 +1,72 @@
+package realm
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+
+	"github.com/10gen/realm-cli/internal/utils/api"
+)
+
+const (
+	exportPathPattern = appPathPattern + "/export"
+
+	exportQueryForSourceControl = "source_control"
+	exportQueryIsTemplated      = "template"
+	exportQueryVersion          = "version"
+
+	mediaParamFilename = "filename"
+
+	trueVal = "true"
+)
+
+// ExportRequest is a Realm application export request
+type ExportRequest struct {
+	ConfigVersion    AppConfigVersion
+	ForSourceControl bool
+	IsTemplated      bool
+}
+
+func (c *client) Export(groupID, appID string, req ExportRequest) (string, io.ReadCloser, error) {
+	options := api.RequestOptions{
+		UseAuth: true,
+		Query: map[string]string{
+			exportQueryVersion: DefaultAppConfigVersion.String(),
+		},
+	}
+
+	if req.ConfigVersion != AppConfigVersionNil {
+		options.Query[exportQueryVersion] = req.ConfigVersion.String()
+	}
+	if req.ForSourceControl {
+		options.Query[exportQueryForSourceControl] = trueVal
+	}
+	if req.IsTemplated {
+		options.Query[exportQueryIsTemplated] = trueVal
+	}
+
+	res, resErr := c.do(http.MethodGet, fmt.Sprintf(exportPathPattern, groupID, appID), options)
+	if resErr != nil {
+		return "", nil, resErr
+	}
+	if res.StatusCode != http.StatusOK {
+		// res.Body.Close()
```

Comment:
so this was carried over from the previous cli, but oddly this was causing an (extra? unrelated?) error: `http: read on closed response body` (I coerced the non-StatusOK response by sending an unsupported config version).

Which makes me want to ask the crowd: when should it be appropriate to call `res.Body.Close()`.  I think ideally its just handled with a single `defer res.Body.Close()` inside the closure of this function, but than forces the client here to perform extra processing on the zipped response.  I don't think the client should also be responsible for writing the zip to disk, but I can't think of other any alternatives to allow us to defer the `res.Body.Close()`

anyone have thoughts here?

Reply:
hmm odd, that likely means something was still trying to read the body after it had been closed (the `UnmarshalServerError` constructor, perhaps?). In this case, that's my guess, and we might just have to close the body _after_ pulling the error out of the response. otherwise, any time you're returning a response body/`ReadCloser`, you're relying on the caller to manage closing out that resource when it's done with it (which seems fine to me). otherwise, a `defer res.Body.Close()` should do the trick so that everything gets cleaned up prior to return.

Reply:
You know I just silently removed the `defer ` when copying this over?  That's totally it, thank you for this.  I'll add the defer back and we should be good to go.  The caller will _still_ be responsible for closing the resource if/when a successful response is returned